lib/manageiq/api/common/application_controller_mixins/exception_handling.rb

```diff
@@ -0,0 +1,43 @@
+module ManageIQ
+  module API
+    module Common
+      module ApplicationControllerMixins
+        module ExceptionHandling
+          DEFAULT_ERROR_CODE = 400
+
+          def self.included(other)
+            other.rescue_from(StandardError, RuntimeError) do |exception|
+              errors, status = collected_errors(exception)
+
+              render :json => errors, :status => status
+            end
+          end
+
+          def collected_errors(exception)
+            errors = []
+            top_level_exception = exception
+
+            until exception.nil?
+              code = exception.respond_to?(:code) ? exception.code : exception.class.to_s
+              errors << {:status => code, :detail => exception.class.to_s + ": " + exception.message}
+              exception = exception.cause
+            end
+
+            # Overwrite the first error since that one most likely doen't have a code
+            errors.first[:status] = error_code_from_class(top_level_exception)
+
+            [{:errors => errors}, errors.first[:status]]
```

Comment:
The status for the document could be different from each of the errors in the body, right?

Reply:
Yes - the status could be different but the goal is that we use the status that comes from the final Exception that was caught here (which we look up the proper code from on L27). It's kinda similar to how `ErrorDocument` works without having to pass that object around after the initial exception. 

One thing I could do is use the `error_code_from_class` method again if you would prefer that, it would be a bit more readable.